src/controllers/krate/publish.rs

```diff
@@ -385,16 +385,40 @@ pub async fn publish(app: AppState, req: BytesRequest) -> AppResult<Json<GoodCra
             // Link this new version to all dependencies
             add_dependencies(conn, &deps, version.id)?;
 
-            // Insert the default version if it doesn't already exist. Compared
-            // to only using a background job, this prevents us from getting
-            // into a situation where a crate exists in the `crates` table but
-            // doesn't have a default version in the `default_versions` table.
-            let inserted_default_versions = diesel::insert_into(default_versions::table)
+            let existing_default_version = default_versions::table
+                .inner_join(versions::table)
+                .filter(default_versions::crate_id.eq(krate.id))
+                .select(DefaultVersion::as_select())
+                .first(conn)
+                .optional()?;
+
+            let published_default_version = DefaultVersion {
+                id: version.id,
+                num: semver,
+                yanked: false,
+            };
+
+            // Upsert the `default_value` determined by the existing `default_value` and the
+            // published version. Note that this could potentially write an outdated version
+            // (although this should not happen regularly), as we might be comparing to an
+            // outdated value.
+            //
+            // Compared to only using a background job, this prevents us from getting into a
+            // situation where a crate exists in the `crates` table but doesn't have a default
+            // version in the `default_versions` table.
+            let default_version = if let Some(existing_default_version) = existing_default_version {
+                std::cmp::max(existing_default_version, published_default_version)
+            } else {
+                published_default_version
+            };
```

Comment:
This could be simplified to the following:
```suggestion
            let default_version = std::cmp::max(Some(published_default_version), existing_default_version)
                .expect("At least the published version should be selected.");
```

Reply:
oops this is an outdated review 😅

Reply:
it's probably correct, but it feels arcane enough to require a comment and proper testing :D